libr/core/tp.c

```diff
@@ -0,0 +1,140 @@
+		char *type = r_anal_type_func_args_type (anal, fcn_name, i);
+		const char *name =r_anal_type_func_args_name (anal, fcn_name, i);
+		const char *place = r_anal_cc_arg (anal, cc, i+1);
+		if (!strcmp (place, "stack")) {
+			for (j = idx; j >= 0; j--) {
+				ut64 write_addr = sdb_num_get (trace, sdb_fmt (-1, "%d.mem.write", j), 0);
+				if (write_addr == sp + size) {
+					ut64 instr_addr = sdb_num_get (trace, sdb_fmt (-1, "%d.addr", j), 0);
+					r_meta_set_string (core->anal, R_META_TYPE_COMMENT, instr_addr,
+						sdb_fmt (-1, "%s %s", type, name));
+					char *tmp = sdb_fmt (-1, "%d.mem.read", j);
+					int i2, array_size = sdb_array_size (trace, tmp);
+					for (i2 = 0; i2 < array_size; i2++) {
+						if (bp_name) {
+							int bp_idx = sdb_array_get_num (trace, tmp, i2, 0) - bp;
+							if ((v =r_anal_var_get (anal, addr, R_ANAL_VAR_KIND_BPV, 1, bp_idx))) {
+								r_anal_var_retype (anal, addr, 1, bp_idx, R_ANAL_VAR_KIND_BPV, type, -1, v->name);
+								r_anal_var_free (v);
+							}
+						}
+						int sp_idx = sp - sdb_array_get_num (trace, tmp, i2, 0);
+						if ((v =r_anal_var_get (anal, addr, R_ANAL_VAR_KIND_SPV, 1, sp_idx))) {
+							r_anal_var_retype (anal, addr, 1, sp_idx, R_ANAL_VAR_KIND_SPV, type, -1, v->name);
+							r_anal_var_free (v);
+						}
+					}
+					break;
+				}
+			}
+			size += r_anal_type_get_size (anal, type) / 8;
+		} else if (!strcmp (place , "stack_rev")) {
+			//TODO
+		} else {
+			for (j = idx; j >= 0; j--) {
+				if (sdb_array_contains (trace, sdb_fmt (-1, "%d.reg.write", j), place, 0)) {
+					ut64 instr_addr = sdb_num_get (trace, sdb_fmt (-1, "%d.addr", j), 0);
+					r_meta_set_string (core->anal, R_META_TYPE_COMMENT, instr_addr,
+						sdb_fmt (-1, "%s %s", type, name));
+					char *tmp = sdb_fmt (-1, "%d.mem.read", j);
+					int i2, array_size = sdb_array_size (trace, tmp);
+					for (i2 = 0; i2 < array_size; i2++) {
+						if (bp_name) {
+							int bp_idx = sdb_array_get_num (trace, tmp, i2, 0) - bp;
+							if ((v =r_anal_var_get (anal, addr, R_ANAL_VAR_KIND_BPV, 1, bp_idx))) {
+								r_anal_var_retype (anal, addr, 1, bp_idx, R_ANAL_VAR_KIND_BPV, type, -1, v->name);
+								r_anal_var_free (v);
+							}
+						}
+						int sp_idx = sp - sdb_array_get_num (trace, tmp, i2, 0);
+						if ((v =r_anal_var_get (anal, addr, R_ANAL_VAR_KIND_SPV, 1, sp_idx))) {
+							r_anal_var_retype (anal, addr, 1, sp_idx, R_ANAL_VAR_KIND_SPV, type, -1, v->name);
+							r_anal_var_free (v);
+						}
+					}
+
+					break;
+				}
+			}
+		}
+		free (type);
+	}
+	//TODO fake emulation of function in to protect stack for being corrupted by
+	//stack cleaning routines just make sure that their is enough space for the
+	//clean instruction to take place
+}
+R_API void r_anal_type_match(RCore *core, RAnalFunction *fcn) {
+	const char *pc = r_reg_get_name (core->anal->reg, R_REG_NAME_PC);
+	ut64 addr = fcn->addr;
+	if (!r_anal_emul_init (core)) {
+		return;
+	}
+	r_core_cmdf (core, "ar PC=0x%"PFMT64x, fcn->addr);
```

Comment:
using api for this specific instruction required lots of initializations that I will make the code more spaghetti